### PR TITLE
Add node_prep script

### PR DESCRIPTION
**Pull request by 83bhp**
Signed-off-by: Pritam Bhavsar <pritam.bhavsar@seagate.com>

### CODE REVIEW

File: srv/components/provisioner/scripts/node_prep.sh [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+    echo "Running prerequisite script" | tee -a ${LOG_FILE}
+    parse_args $@
+    if [[ ! -z $user_name ]]; then
+        if [[ -z $uid -o -z $group -o $uid -o $gid ]]; then
+            echo "Please provide uid, gid & group" | tee -a ${LOG_FILE}
+            exit 1
+        fi
+        create_factory_user $user_name $uid $group $gid
+    fi
+    if [[ ! -z $repo_url ]]; then
```
Comment (83bhp) [inline srv/components/provisioner/scripts/node_prep.sh:218]: Read it from mounted iso path

File: srv/components/provisioner/scripts/node_prep.sh [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+            echo "Package $pkg is already installed."
+    done
+}
+
+
+main()
+{
+    echo "Running prerequisite script" | tee -a ${LOG_FILE}
+    parse_args $@
+    if [[ ! -z $user_name ]]; then
```
Comment (83bhp) [inline srv/components/provisioner/scripts/node_prep.sh:211]: use default user nodeadmin, support.

File: srv/components/provisioner/scripts/node_prep.sh [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+PRVSNR_ROOT="/opt/seagate/cortx/provisioner"
+localhost_ip="127.0.0.1"
+minion_id="srvnode-0"
+user_name=
+uid=
+group=
+gid=
+repo_url=
+
+
```
Comment (83bhp) [inline srv/components/provisioner/scripts/node_prep.sh:32]: add usage

File: srv/components/provisioner/scripts/node_prep.sh [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+#
+
```
Comment (83bhp) [inline srv/components/provisioner/scripts/node_prep.sh:18]: install it at /opt/seagate/cortx/startup.sh

File: srv/components/provisioner/scripts/node_prep.sh [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+    fi
+
+    minion_file="${PRVSNR_ROOT}/srv/components/provisioner/salt_master/files/master"
+    master_file="${PRVSNR_ROOT}/srv/components/provisioner/salt_minion/files/minion"
+
+    yes | cp -f "${master_file}" /etc/salt/master
+    yes | cp -f "${minion_file}" /etc/salt/minion
+
+    local line_to_replace=$(grep -m1 -noP "master: " /etc/salt/minion|tail -1|cut -d: -f1)
+    echo "Setting local-master in /etc/salt/minon" | tee -a ${LOG_FILE}
```
Comment (83bhp) [inline srv/components/provisioner/scripts/node_prep.sh:101]: make this debug log.

File: srv/components/provisioner/scripts/node_prep.sh [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+    yes | cp -f /etc/salt/minion /etc/salt/minion.bkp
+    sed -i "${line_to_replace}s|^master:.*|master: ${localhost_ip}|" /etc/salt/minion
+    echo "Done" | tee -a ${LOG_FILE}
+
+    echo $minion_id > /etc/salt/minion_id
+
+    echo "Restarting salt services" | tee -a ${LOG_FILE}
```
Comment (83bhp) [inline srv/components/provisioner/scripts/node_prep.sh:108]: make the display output generic, no salt/provisioner keywords

File: srv/components/provisioner/scripts/cortx_prep.sh [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+        echo "Preparing the Cortx ConfStore with default configuration" | tee -a ${LOG_FILE}
+        cortx_setup prepare_confstore
+    else
+        echo "WARNING: cortx_setup commands are not installed" | tee -a ${LOG_FILE}
+        echo "Please install the cortx_setup commands and run: 'cortx_setup prepare_confstore' manually"
+    fi
+}
+
+main $@
+echo "Done"
```
Comment (nitin-seagate) [inline srv/components/provisioner/scripts/cortx_prep.sh:291]: Add new line to end of file...

File: srv/components/provisioner/scripts/cortx_prep.sh [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+        echo "This is VM" >> ${LOG_FILE}
+        if [[ ! -z $repo_url ]]; then
+            setup_repos $repo_url
+        else
+            echo "ERROR: target-build not provided, it's mandetary option for VM"
+            echo "ERROR: Please provide the target-build " | tee -a ${LOG_FILE}
+            exit 1
+        fi
+    fi
+    install_cortx_pkgs
```
Comment (nitin-seagate) [inline srv/components/provisioner/scripts/cortx_prep.sh:277]: Lets write another function to handle non-cortx packages installation

File: srv/components/provisioner/scripts/cortx_prep.sh [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+            echo "ERROR: Provisioner configuration manager failed to connect to local server" | tee -a ${LOG_FILE}
+            exit 1
+        fi
+        try=$(( $try + 1 ))
+    done
+    echo "Minion on Server A started successfully" >> ${LOG_FILE}
+    echo "Done" | tee -a ${LOG_FILE}
+}
+
+create_factory_user()
```
Comment (nitin-seagate) [inline srv/components/provisioner/scripts/cortx_prep.sh:133]: Remove this function, as we plan to create users as part of node finalize step towards the end of manufacturing...

File: srv/components/provisioner/scripts/cortx_prep.sh [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+            -h|--help)
+                usage; exit 0
+            *) echo "Invalid option $1"; usage; exit 1;;
+        esac
+    done
+}
+
+config_local_salt()
+{
+
```
Comment (nitin-seagate) [inline srv/components/provisioner/scripts/cortx_prep.sh:71]: Remove extra line

File: srv/components/provisioner/scripts/cortx_prep.sh [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+localhost_ip="127.0.0.1"
+minion_id="srvnode-0"
+user_name=
+uid=
+group=
+gid=
+repo_url=
+ISO_CORTX_PATH="/opt/isos/cortx-*-single.iso"
+ISO_OS_PATH="/opt/isos/cortx-os-*.iso"
+nodejs_tar="http://cortx-storage.colo.seagate.com/releases/cortx/github/integration-custom-ci/centos-7.8.2003/custom-build-1969/3rd_party/commons/node/node-v12.13.0-linux-x64.tar.xz"
```
Comment (nitin-seagate) [inline srv/components/provisioner/scripts/cortx_prep.sh:33]: Can this be handled in separate function to install non-cortx packages

File: srv/components/provisioner/scripts/cortx_prep.sh [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+    useradd -c "$user_suport" -d /home/${user_support}
+    echo "Done" | tee -a ${LOG_FILE}
+
+}
+
+setup_repos()
+{
+    #repo="http://cortx-storage.colo.seagate.com/releases/cortx/github/main/centos-7.8.2003/1291/prod/"
+
+    repo=$1
```
Comment (nitin-seagate) [inline srv/components/provisioner/scripts/cortx_prep.sh:154]: Can you confirm if the argument is correct... Guess it should be `$repo_url`

File: srv/components/provisioner/scripts/cortx_prep.sh [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+#
+
+set -euE
+export LOG_FILE="${LOG_FILE:-/var/log/seagate/provisioner/node_prep.log}"
+mkdir -p $(dirname "${LOG_FILE}")
+
+PRVSNR_ROOT="/opt/seagate/cortx/provisioner"
+localhost_ip="127.0.0.1"
+minion_id="srvnode-0"
+user_name=
```
Comment (nitin-seagate) [inline srv/components/provisioner/scripts/cortx_prep.sh:26]: Remove unused:
`user_name, uid, group, gid`

File: srv/components/provisioner/scripts/cortx_prep.sh [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+uid=
+group=
+gid=
+repo_url=
```
Comment (nitin-seagate) [inline srv/components/provisioner/scripts/cortx_prep.sh:30]: Set default value as some sample URL location

File: srv/components/provisioner/scripts/cortx_prep.sh [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+        "udx-discovery.x86_64"
+    )
+
+    for pkg in ${cortx_pkgs[@]}; do
+        echo "Installing $pkg" | tee -a ${LOG_FILE}
+        rpm --quiet -qi $pkg ||  yum install --nogpgcheck -y -q $pkg  && \
+            echo "Package $pkg is already installed."
+    done
+
+    echo "Installing nodejs" | tee -a ${LOG_FILE}
```
Comment (nitin-seagate) [inline srv/components/provisioner/scripts/cortx_prep.sh:205]: Lets move this logic of installation nodejs as separate function to handle `install_noncortx_pkgs`

File: srv/components/provisioner/scripts/cortx_prep.sh [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+    echo "Downloading nodeja tarball" >> ${LOG_FILE}
+    wget ${nodejs_tar}
+    echo "Extracting the tarball" >> ${LOG_FILE}
+    mkdir /opt/nodejs
+    tar -C /opt/nodejs/ -xvf node-v12.13.0-linux-x64.tar.xz
+    echo "The extrracted tarball is kept at /opt/nodejs, removing the tarball ${nodejs_tar}" >> ${LOG_FILE}
```
Comment (nitin-seagate) [inline srv/components/provisioner/scripts/cortx_prep.sh:211]: Make this debug log, fix typo spelling errors

File: srv/components/provisioner/scripts/cortx_prep.sh [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+    )
+
+    for pkg in ${cortx_pkgs[@]}; do
+        echo "Installing $pkg" | tee -a ${LOG_FILE}
+        rpm --quiet -qi $pkg ||  yum install --nogpgcheck -y -q $pkg  && \
+            echo "Package $pkg is already installed."
+    done
+
+    echo "Installing nodejs" | tee -a ${LOG_FILE}
+    echo "Downloading nodeja tarball" >> ${LOG_FILE}
```
Comment (nitin-seagate) [inline srv/components/provisioner/scripts/cortx_prep.sh:206]: Make this debug log, fix typo spelling errors

File: srv/components/provisioner/scripts/cortx_prep.sh [hunk 18]
```diff
@@ -0,0 +1,291 @@
...
+    until salt -t 1 $minion_id test.ping >/dev/null 2>&1
+    do
+        if [[ "$try" -gt "$max_tries" ]]; then
+            echo "ERROR: minion srvnode-0 seems still not ready after $max_tries attempts." >> ${LOG_FILE}
+            echo "ERROR: Provisioner configuration manager failed to connect to local server" | tee -a ${LOG_FILE}
+            exit 1
+        fi
+        try=$(( $try + 1 ))
+    done
+    echo "Minion on Server A started successfully" >> ${LOG_FILE}
```
Comment (nitin-seagate) [inline srv/components/provisioner/scripts/cortx_prep.sh:129]: Change this to `Minion on $minion_id started successfully`

File: srv/components/provisioner/scripts/cortx_prep.sh [hunk 19]
```diff
@@ -0,0 +1,291 @@
...
+    echo "Accepting the salt key for minion $minion_id" >> ${LOG_FILE}
+    salt-key -y -a $minion_id
+
+    # Check if salt '*' test.ping works
+    echo "Waiting for Provisioner configuratoin manager to become ready" | tee -a ${LOG_FILE}
+    try=1; max_tries=10
+    until salt -t 1 $minion_id test.ping >/dev/null 2>&1
+    do
+        if [[ "$try" -gt "$max_tries" ]]; then
+            echo "ERROR: minion srvnode-0 seems still not ready after $max_tries attempts." >> ${LOG_FILE}
```
Comment (nitin-seagate) [inline srv/components/provisioner/scripts/cortx_prep.sh:123]: Change this to `Minion on $minion_id seems still not ready after $max_tries attempts.`

File: srv/components/provisioner/scripts/cortx_prep.sh [hunk 20]
```diff
@@ -0,0 +1,291 @@
...
+    systemctl start salt-master
+    systemctl restart salt-minion
+    echo "Done" | tee -a ${LOG_FILE}
+
+    echo "INFO: Waiting for key of $minion_id to become connected to salt-master" >> ${LOG_FILE}
+    try=1; max_tries=10
+    until salt-key --list-all | grep "srvnode-0" >/dev/null 2>&1
+    do
+        if [[ "$try" -gt "$max_tries" ]]; then
+            echo "ERROR: Key for salt-minion $minion_id not listed after $max_tries attemps." >> ${LOG_FILE}
```
Comment (nitin-seagate) [inline srv/components/provisioner/scripts/cortx_prep.sh:105]: Remove keyword `salt-minion`

**Review by nitin-seagate — CHANGES_REQUESTED** [review]

File: srv/components/provisioner/scripts/cortx_prep.sh (hunk 7 above)
Comment (83bhp) [reply srv/components/provisioner/scripts/cortx_prep.sh:291]: done

File: srv/components/provisioner/scripts/cortx_prep.sh (hunk 9 above)
Comment (83bhp) [reply srv/components/provisioner/scripts/cortx_prep.sh:133]: removed

File: srv/components/provisioner/scripts/cortx_prep.sh (hunk 11 above)
Comment (83bhp) [reply srv/components/provisioner/scripts/cortx_prep.sh:33]: Done

File: srv/components/provisioner/scripts/cortx_prep.sh (hunk 16 above)
Comment (83bhp) [reply srv/components/provisioner/scripts/cortx_prep.sh:211]: It's debug only, redirected to LOG_FILE directly, added 'DEBUG' keyword now

File: srv/components/provisioner/scripts/cortx_prep.sh (hunk 18 above)
Comment (83bhp) [reply srv/components/provisioner/scripts/cortx_prep.sh:129]: Corrected

File: srv/components/provisioner/scripts/cortx_prep.sh (hunk 19 above)
Comment (83bhp) [reply srv/components/provisioner/scripts/cortx_prep.sh:123]: done

File: srv/components/provisioner/scripts/cortx_prep.sh (hunk 20 above)
Comment (83bhp) [reply srv/components/provisioner/scripts/cortx_prep.sh:105]: It's debug log, redirected to LOG_FILE directly, added 'DEBUG' keyword now

File: srv/components/provisioner/scripts/cortx_prep.sh [hunk 21]
```diff
@@ -0,0 +1,291 @@
...
+        case $1 in
+            -t|--target-build)
+                if [[ -z "$2" ]]; then
+                    echo "Error: URL for target cortx build not provided" && usage && exit 1;
+                fi
+                repo_url=$2
+                shift 2
+                ;;
+            -h|--help)
+                usage; exit 0
```
Comment (nitin-seagate) [inline srv/components/provisioner/scripts/cortx_prep.sh:63]: Check this will result in error: Correct this one
`usage; exit 0;;`

**Review by nitin-seagate — CHANGES_REQUESTED** [review]

**Comment by 83bhp** [comment]
Closing this PR as alternative PR has been raised:
https://github.com/Seagate/cortx-prvsnr/pull/1426

File: srv/components/provisioner/scripts/cortx_prep.sh (hunk 13 above)
Comment (83bhp) [reply srv/components/provisioner/scripts/cortx_prep.sh:26]: done

File: srv/components/provisioner/scripts/cortx_prep.sh (hunk 14 above)
Comment (83bhp) [reply srv/components/provisioner/scripts/cortx_prep.sh:30]: It's intentional so as to be able check whether the --target-build <url> options is provided by user or not.

File: srv/components/provisioner/scripts/cortx_prep.sh (hunk 21 above)
Comment (83bhp) [reply srv/components/provisioner/scripts/cortx_prep.sh:63]: Thanks, fixed!!

File: srv/components/provisioner/scripts/cortx_prep.sh (hunk 12 above)
Comment (83bhp) [reply srv/components/provisioner/scripts/cortx_prep.sh:136]: yes, it's correct. This is local variable to setup_repos()

File: srv/components/provisioner/scripts/cortx_prep.sh (hunk 8 above)
Comment (83bhp) [reply srv/components/provisioner/scripts/cortx_prep.sh:275]: yes, moved the non-cortx package installation to another function - install_dependency_pkgs, check the line above (L274)

**Review by 83bhp — COMMENTED** [review]
Addressed all the comments, please check.

File: srv/components/provisioner/scripts/cortx_prep.sh [hunk 22]
```diff
@@ -0,0 +1,306 @@
...
+            echo "ERROR: Please provide the target-build " | tee -a ${LOG_FILE}
+            exit 1
+        fi
+    fi
+
+    install_dependency_pkgs
+    install_cortx_pkgs
+    config_local_salt
+
+
```
Comment (nitin-seagate) [inline srv/components/provisioner/scripts/cortx_prep.sh:299]: Just after configuring local salt please and before prepare **ConfStore** execute this command:
`components.provisioner.config.machine_id.reset`

**Review by nitin-seagate — APPROVED** [review]
Minor changes requested...

**Review by nitin-seagate — APPROVED** [review]
Looks good to me

**Review by sradhanand-pati — APPROVED** [review]

**Comment by 83bhp** [comment]
Closing this as the alternative PR was raised to address cla-signed issue:
https://github.com/Seagate/cortx-prvsnr/pull/1437

**Pull request closed without merging**